server/Sample/sfreerdp.h

```diff
@@ -60,6 +60,10 @@ struct test_peer_context
 	HANDLE debug_channel_thread;
 #if defined(CHANNEL_AUDIN_SERVER)
 	audin_server_context* audin;
+	AUDIO_FORMAT* audin_server_formats;
```

Comment:
on a second look, might these be moved to `audin_server_context`? they should be required by every server implementation and their scope is the same, so why not unify?

Reply:
Their allocation can be static (I do that in g-r-d). And in that case, they have to be always set to `NULL` before clearing the `audin_server_context` to prevent a crash.

The dummy handling in the test servers here in FreeRDP unfortunately use the same struct for the audin handling, that is also used for anything else, so this can be a bit perceived to be annoying here in the test servers.
However, normally the audin handling (or any other channel) is usually handled in its own source file with its own struct, where only the related settings are (e.g. PipeWire or Pulseaudio hook-up, etc.).

Reply:
@pnowack I was more thinking along having some of the stuff moved to the channel. (you need to copy the formats initially, correct)
it would then be possible to have `AudinServerReceiveVersion`, `AudinServerReceiveFormats`, `AudinServerReceiveFormatChange` to have a common implementation.
(`	audin->serverVersion = SNDIN_VERSION_Version_2;` should also be moved to init code, so you are ready to go with just the remaining callbacks you need)

Reply:
> (you need to copy the formats initially, correct)

Not really, I just assign the pointer to the struct, containing my formats: https://gitlab.gnome.org/pnowack/gnome-remote-desktop/-/blob/78e2ef9194a3cad196a0f6b4920a7e0606157b83/src/grd-rdp-audio-input.c#L257

> (	audin->serverVersion = SNDIN_VERSION_Version_2; should also be moved to init code, so you are ready to go with just the remaining callbacks you need)

We don't know how the audin protocol changes in the future. I would like to preserve the behaviour, where the server can choose what protocol version it wants to support.
Would it be okay for you, if we set a default value in `audin_server_context_new()`? Then the server implementation can override that value, if it wants to?

`AudinServerReceiveFormatChange` cannot be in a common implementation. In version 2 of the protocol, the server side can tell the client to use a different format.
That format change does not apply immediately. The client has to confirm that change by sending that PDU itself to server side too.
After receiving that PDU, the server can then (and must) expect the (encoded) samples to be in the new format, but not before that.

Reply:
> > (you need to copy the formats initially, correct)
> 
> Not really, I just assign the pointer to the struct, containing my formats: https://gitlab.gnome.org/pnowack/gnome-remote-desktop/-/blob/78e2ef9194a3cad196a0f6b4920a7e0606157b83/src/grd-rdp-audio-input.c#L257
> 
yes, that is what you do now. if the behaviour is changed to you provide this pointer (and size) to the `audin` context and it creates a copy on `init` then my suggestion would work. (more to do, but only once for each server implementation)

> > (	audin->serverVersion = SNDIN_VERSION_Version_2; should also be moved to init code, so you are ready to go with just the remaining callbacks you need)
> 
> We don't know how the audin protocol changes in the future. I would like to preserve the behaviour, where the server can choose what protocol version it wants to support. Would it be okay for you, if we set a default value in `audin_server_context_new()`? Then the server implementation can override that value, if it wants to?
> 
you can always check for the version, talking about defaults here.

> > AudinServerReceiveFormatChange to have a common implementation
> 
> `AudinServerReceiveFormatChange` cannot be in a common implementation. In version 2 of the protocol, the server side can tell the client to use a different format. That format change does not apply immediately. The client has to confirm that change by sending that PDU itself to server side too. After receiving that PDU, the server can then (and must) expect the (encoded) samples to be in the new format, but not before that.
if all the required variables are part of the `audin` context, then no, you can just let it be handled there. (that would be true for the shadow and sample server, you can always provide your own specific implementation if you want more)

Reply:
If I think a bit more about it, `AudinServerIncomingData` could also get a `default` implementation (with some simple basic measuring algorithm)